### PR TITLE
Add option to plot utilities to generate PNG output rather than inline plots

### PR DESCRIPTION
**Pull request by mnlevy1981**
This PR will generate images and metadata for @andersy005's dashboard. The first commit produces a separate JSON file for each generated plot, but @andersy005 will follow behind with a tool to clean that all up :) Additional commits are needed to generate PNG files from other plot types (time series, etc).

### CODE REVIEW

File: notebooks/utils/utils.py [hunk 1]
```diff
@@ -212,3 +215,29 @@ def print_key_metadata(ds, msg=None):
...
+    root_dir = pathlib.Path(root_dir)
+    files = sorted(root_dir.rglob(f"**/*{extension}"))
+    data = []
+    if files:
+        for file in files:
+            data.append(json.load(file.open()))
+        return pd.DataFrame(data)
+    else:
+        print(f"Found 0 files with extension={extension} in {root_dir}.")
+        return pd.DataFrame()
```
Comment (andersy005) [inline notebooks/utils/utils.py:243]: @mnlevy1981, so far this function returns the following

```python
In [1]: from utils import generate_plot_catalog

In [2]: root_dir = "/glade/work/mlevy/codes/hi-res_analysis/notebooks/images"

In [5]: df = generate_plot_catalog(root_dir)

In [6]: df.head()
Out[6]: 
                                            filepath  apply_log10        date          varname     plot_type
0  images/g.e22.G1850ECO_JRA_HR.TL319_t13.003/CaC...        False  0001-01-16  CaCO3_FLUX_100m  summary_plot
1  images/g.e22.G1850ECO_JRA_HR.TL319_t13.003/CaC...         True  0001-01-16  CaCO3_FLUX_100m  summary_plot
2  images/g.e22.G1850ECO_JRA_HR.TL319_t13.003/CaC...        False  0001-02-15  CaCO3_FLUX_100m  summary_plot
3  images/g.e22.G1850ECO_JRA_HR.TL319_t13.003/CaC...         True  0001-02-15  CaCO3_FLUX_100m  summary_plot
4  images/g.e22.G1850ECO_JRA_HR.TL319_t13.003/CaC...        False  0001-03-16  CaCO3_FLUX_100m  summary_plot
```

I haven't decided what to do with the filepaths yet

File: notebooks/utils/utils.py [hunk 2]
```diff
@@ -212,3 +215,35 @@ def print_key_metadata(ds, msg=None):
...
+    root_dir = pathlib.Path(root_dir)
+    image_dir = root_dir / image_dir_name
+    image_dir.exists()
+    files = sorted(root_dir.rglob(f"**/*{extension}"))
+    data = []
+    if files:
+        for file in files:
+            metadata = json.load(file.open())
+            metadata["filepath"] = (
+                (root_dir / metadata["filepath"]).absolute().as_posix()
```
Comment (andersy005) [inline notebooks/utils/utils.py:243]: The latest commit uses absolute paths for images

**Comment by mnlevy1981** [comment]
I'm going to turn off draft status because at this point it looks `Sanity Check.ipynb` is the only notebook with inline plots still, and it's only 133 KB on disk. I've verified that the number of plots I'm creating matches the number of plots being removed from each of the following notebooks (comments in the code block indicate how many `png` files exist in each directory; totals should sum to the number of plots removed when diffing the notebook against `master`):

* `plot_suite_003.ipynb`

   ```
   # histogram/g.e22.G1850ECO_JRA_HR.TL319_t13.003: 160
   # time_series/g.e22.G1850ECO_JRA_HR.TL319_t13.003: 24
   $ git diff master plot_suite_003.ipynb | grep ^-.*png | wc -l
   184
   ```

* `plot_suite_004.ipynb`

   ```
   # histogram/g.e22.G1850ECO_JRA_HR.TL319_t13.003: 680
   # time_series/g.e22.G1850ECO_JRA_HR.TL319_t13.003: 24
   $ git diff master plot_suite_004.ipynb | grep ^-.*png | wc -l
   704
   ```

* `plot_suite_1deg.ipynb`

   ```
   # histogram/g.e22b05.G1850ECOIAF_JRA.TL319_g17.cocco.001: 400
   # time_series/g.e22b05.G1850ECOIAF_JRA.TL319_g17.cocco.001: 24
   $ git diff master plot_suite_1deg.ipynb | grep ^-.*png | wc -l
   184
   ```

* `plot_suite_maps_0001_003.ipynb`

   ```
   # summary_map/g.e22.G1850ECO_JRA_HR.TL319_t13.003: 480
   $ git diff master plot_suite_maps_0001_003.ipynb | grep ^-.*png | wc -l
   480
   ```

* `plot_suite_maps_0001_004.ipynb`

   ```
   # summary_map/g.e22.G1850ECO_JRA_HR.TL319_t13.004: 480
   $ git diff master plot_suite_maps_0001_004.ipynb | grep ^-.*png | wc -l
   480
   ```

* `plot_suite_maps_0095_1deg.ipynb`

   ```
   # summary_map/g.e22b05.G1850ECOIAF_JRA.TL319_g17.cocco.001: 480
   $ git diff master plot_suite_maps_0095_1deg.ipynb | grep ^-.*png | wc -l
   480
   ```

* `trend_maps.003.ipynb`

   ```
   # trend_hist/g.e22.G1850ECO_JRA_HR.TL319_t13.003: 30
   # trend_map/g.e22.G1850ECO_JRA_HR.TL319_t13.003: 30
   $ git diff trend_maps.003.ipynb | grep ^-.*png | wc -l
   60
   ```

* `trend_maps.004.ipynb`

   ```
   # trend_hist/g.e22.G1850ECO_JRA_HR.TL319_t13.004: 30
   # trend_map/g.e22.G1850ECO_JRA_HR.TL319_t13.004: 30
   $ git diff trend_maps.004.ipynb | grep ^-.*png | wc -l
   60
   ```

****

Still, I think there are several open questions we should discuss before merging:

1. File naming conventions: for now, running any of the notebooks that produce png output will create an `images/` directory. Inside that directory, there are subdirectories for each type of plot (`histogram/`, `summary_map/`, `time_series/`, `trend_hist/`, and `trend_map/`). Each of those contains subdirectories for each of the cases, and then there is a pattern defined for each type of plot in [utils/PlotTypeClass.py](https://github.com/mnlevy1981/HiRes-CESM-analysis/blob/generate_pngs_for_dashboard/notebooks/utils/PlotTypeClass.py). Subdirectories for each case is not necessary, and I'm happy to just include the case name in the file name instead
2. Metadata conventions

   * As a first pass, I am just creating one `json` file per plot... so the plot `plot_type/case/filename.png` is associated with `plot_type/case/metadata/filename.json`. I haven't tied in Anderson's `generate_plot_catalog()` function yet
   * We should look closely at what is currently going into the json files and make sure I'm not omitting anything

3. Should we do some refactoring to bring routines from `Plotting.py` into `PlotTypeClass.py`? Perhaps the notebooks should be importing `PlotTypeClass` and the plots themselves can be generated by class method objects?

   * Perhaps we would have an additional level of inheritance in `PlotTypeClass` classes... rather than `_PlotTypeBaseClass -> SummaryMapClass & TrendMapClass` we could have `_PlotTypeBaseClass -> _MapBaseClass -> SummaryMapClass & TrendMapClass` and have the actual plotting be defined in `_MapBaseClass`? I haven't actually sat down and thought about what this would look like, but it seems like `SummaryMapClass` and `TrendMapClass` would have more in common with each other than, say `TrendMapClass` and `TrendHistClass`

**Comment by mnlevy1981** [comment]
@andersy005 -- I should have mentioned this in the previous comment, but png files are in `/glade/work/mlevy/hi-res_BGC_JRA/plots_from_analysis/` if you want to play with those in your dashboard; I think the best way to figure out what metadata is missing is to look at a mock-up and see what we wish we could sort by :)

**Comment by mnlevy1981** [comment]
Also on the to-do list: I need to add an argument to allow control over the figure size / resolution when writing to disk.

**Comment by klindsay28** [comment]
I suggest that the `__init__` and `get_filepaths` methods in class `_PlotTypeBaseClass` have `*args` and `**kwargs` arguments.

That way, calls to these methods, that should never be called, will actually get to the `raise NotImplementedError` statements.

As is, a user of a derived class that doesn't implement these methods will get errors about unexpected keyword arguments or too many positional arguments when the `__init__` and `get_filepaths` methods in `_PlotTypeBaseClass` are called (, assuming that the user passes arguments to the methods). These argument error messages aren't as helpful as the message in the `raise NotImplementedError` statements.

**Comment by mnlevy1981** [comment]
eefc145 introduces a few updates that are mostly related:

```
 A few changes to plot generation:

1. Address Keith's comment about *args and **kwargs
2. Use time_bounds to determine interval for histograms and time series
3. Reorganize output so that case name is the top-level directory rather than
   plot type -- this will get clunky when comparing multiple cases, but it
   lines up better with the directory structure of webext.cgd.ucar.edu. When we
   have plots comparing multiple cases, we can figure out a good naming scheme
4. generate_plot_catalog() can print relative paths (and the filepath in the
   json file is relative to the casename rather than the notebooks directory)
```

The first bullet point is just some low-hanging fruit; the rest ties into making data available for the dashboard. I've started posting some PNG files (and associated CSV catalogs) on [webext](https://webext.cgd.ucar.edu/g.e22.G1850ECO_JRA_HR.TL319_t13.004/); so now we can inline images directly from that server:

POC Flux at 100m timeseries from 004

Also, @mgrover1 pointed out that we should use `time_bounds` instead of `time` to determine the beginning and end of the interval -- that's why the image above is [correctly] labelled from 0001-01-01 to 0017-12-31 where previous versions claimed it was a time series from 0001-01-16 to 0017-12-16

I'm waiting for one more notebook to finish running, then I'll make a commit with those updates.

**Comment by mnlevy1981** [comment]
I noticed that [`png_catalog.csv`](https://webext.cgd.ucar.edu/g.e22.G1850ECO_JRA_HR.TL319_t13.004/png_catalog.csv) has the index in the first column... but when I set `index_label=False` it doesn't go away. I'll play with it more during the week.

**Comment by mnlevy1981** [comment]
From chatting with @mgrover1, it looks like we pull useful information from `isel_dict` for the filename (e.g. `basins--Pacific` and `z_t--35109.35`) but the metadata contains entry like `"isel_dict": {"basins": 0}` and `"isel_dict": {"z_t": 0}`. Those should be replaced with the actual dimension values... and maybe instead of `isel_dict` I can call it `dim_reduction` or something?

**Comment by mnlevy1981** [comment]
@mgrover1 and @klindsay28  -- this is ready for review.

@mgrover1 -- I've moved the latest output to webext, hopefully the dashboard reflects the change from `isel_dict` to `reduced_dims`.

@klindsay28 -- for now, I've left `isel_dict` in `diag_metadata.yaml`; we talked earlier this week about changing that so (for example) that the code uses `sel_dict(z_t=35000, method="nearest")` instead of `isel_dict(z_t=28)` but I'd rather address that in a separate issue / PR.

I believe all the other comments mentioned in meetings / this PR thread have been addressed.

**Pull request merged**